openmdao/core/problem.py

```diff
@@ -531,11 +567,170 @@ def run(self):
 
         return self.run_driver()
 
+    def _setup_recording(self):
+        """
+        Set up case recording.
+        """
+        model = self.model
+        driver = self.driver
+
+        mydesvars = myobjectives = myconstraints = myresponses = set()
+
+        incl = self.recording_options['includes']
+        excl = self.recording_options['excludes']
+
+        rec_desvars = self.recording_options['record_desvars']
+        rec_objectives = self.recording_options['record_objectives']
+        rec_constraints = self.recording_options['record_constraints']
+
+        all_desvars = {n for n in driver._designvars
+                       if check_path(n, incl, excl, True)}
+        all_objectives = {n for n in driver._objs
+                          if check_path(n, incl, excl, True)}
+        all_constraints = {n for n in driver._cons
+                           if check_path(n, incl, excl, True)}
+        if rec_desvars:
+            mydesvars = all_desvars
+
+        if rec_objectives:
+            myobjectives = all_objectives
+
+        if rec_constraints:
+            myconstraints = all_constraints
+
+        # get the includes that were requested for this Driver recording
+        if incl:
+            # The my* variables are sets
+
+            # First gather all of the desired outputs
+            # The following might only be the local vars if MPI
+            mysystem_outputs = {n for n in model._outputs
+                                if check_path(n, incl, excl)}
+
+            # If MPI, and on rank 0, need to gather up all the variables
+            #    even those not local to rank 0
+            if MPI:
+                all_vars = model.comm.gather(mysystem_outputs, root=0)
+                if MPI.COMM_WORLD.rank == 0:
+                    mysystem_outputs = all_vars[-1]
+                    for d in all_vars[:-1]:
+                        mysystem_outputs.update(d)
+
+            # de-duplicate mysystem_outputs
+            mysystem_outputs = mysystem_outputs.difference(all_desvars, all_objectives,
+                                                           all_constraints)
+
+        # if rec_inputs:
+        #     root = self.model
+        #     myinputs = {n for n in root._inputs
+        #                 if check_path(n, incl, excl)}
+        #
+        #     if MPI:
+        #         all_vars = root.comm.gather(myinputs, root=0)
+        #         if MPI.COMM_WORLD.rank == 0:
+        #             myinputs = all_vars[-1]
+        #             for d in all_vars[:-1]:
+        #                 myinputs.update(d)
+
+        if MPI:  # filter based on who owns the variables
+            # TODO Eventually, we think we can get rid of this next check. But to be safe,
+            #       we are leaving it in there.
+            if not model.is_active():
```

Comment:
Can prob.root ever not be active?

Reply:
From Bret: Currently prob.root can't be inactive